Add JSON RPC call to send button press event.

bendavid opened the pull request:
## Description
Adds Input.ButtonEvent method to JSON RPC api to send button presses to input manager.

## Motivation and Context
Using the existing Input.ExecuteAction method bypasses the window-id based mapping which is used for processing button events from the EventServer.  This allows to reproduce the EventServer or remote/keyboard behaviour using a JSON RPC call, avoiding the need to implement both the JSON RPC and Event Server API in clients which would like to send button presses but also use other JSON RPC functionality.

See e.g. discussion in 
https://forum.kodi.tv/showthread.php?tid=348835&pid=2896694#pid2896694

Tested using Home Assistant Kodi integration with service call parameters like
('kodi', 'call_method', { 'entity_id': kodi_entity, 'method' : 'Input.Button', 'button' : 'up', 'keymap' : 'R1'})

bendavid committed:
Add JSON RPC call to send button press event and move button string
translation to common place.

rmrector commented:
I dig it. Can you add an optional int param to pass in hold time? Default to zero like you have hardcoded now.

And the [JSON-RPC version](https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/version.txt) will need to be increased. It's a non-breaking feature addition so increase the second number and change the third back to 0 (right now the result would be 10.6.0).

garbear commented:
nice feature! I would add support for hold time, and you have two options:

* Per @rmrector, add a holdtime parameter and pass through the `XBMC_BUTTON` event
* Use two events, one for button down, one for button up

The downside of passing a "hold time" is that you have to maintain and communicate state. Passing separate "down" and "up" events lets Kodi hold that state for you. This keeps clients as simple as possible.

Beyond state ownership, the other impact is retro games - games require a "button up" event, because ROMs *must* store state. If you know my history, you'll know I favor the approach that makes games work with no extra effort :)

garbear commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  std::string keymap = parameterObject["keymap"].asString();
+
+  int keycode = 0;
+  if ( keymap.compare("KB") == 0 ) // standard keyboard map
```
you can use `operator==()` e.g. `keymap == "KB"` for the same effect with greater code clarity

bendavid replied:
This code was actually copied from
https://github.com/xbmc/xbmc/blob/6bc162afc45fb7054d02dedb57fc4aa986f5e7cd/xbmc/network/EventClient.cpp#L62-L88

Let me think about moving it to some suitable common place as well.

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
   return SendAction(action);
 }
 
+JSONRPC_STATUS CInputOperations::Button(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  int keycode = 0;
+  if ( keymap.compare("KB") == 0 ) // standard keyboard map
+  {
+  keycode = CKeyboardTranslator::TranslateString( button.c_str() );
```
> Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

`CKeyboardTranslator::TranslateString` takes a `const std::string&` as argument, so there is no need to call `.c_str()`.

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  }
+  else if  ( keymap.compare("XG") == 0 ) // xbox gamepad map
+  {
+  keycode = CGamepadTranslator::TranslateString( button.c_str() );
```
> Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

`CKeyboardTranslator::TranslateString` takes a `std::string` as argument, so there is no need to call `.c_str()`.

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  }
+  else if  ( keymap.compare("R1") == 0 ) // xbox remote map
+  {
+  keycode = CIRTranslator::TranslateString( button.c_str() );
```
> Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

`CKeyboardTranslator::TranslateString` takes a `std::string` as argument, so there is no need to call `.c_str()`.

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  }
+  else if  ( keymap.compare("R2") == 0 ) // xbox universal remote map
+  {
+  keycode = CIRTranslator::TranslateUniversalRemoteString( button.c_str() );
```
> Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

`CKeyboardTranslator::TranslateString` takes a `const std::string&` as argument, so there is no need to call `.c_str()`.

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  {
+    return InvalidParams;
+  }
+  
```
remove the whitespaces in this line

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  newEvent.type = XBMC_BUTTON;
+  newEvent.keybutton.button = keycode;
+  newEvent.keybutton.holdtime = 0;
+  
```
remove the whitespaces in this line

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  newEvent.keybutton.button = keycode;
+  newEvent.keybutton.holdtime = 0;
+  
+  CInputManager &inputmanager = CServiceBroker::GetInputManager();
```
> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +88,45 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+  inputmanager.OnEvent(newEvent);
+  
+  return ACK;
+
```
remove the empty line at the end of the function.

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.h:
```diff
@@ -19,6 +19,7 @@ namespace JSONRPC
   public:
     static JSONRPC_STATUS SendText(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result);
     static JSONRPC_STATUS ExecuteAction(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result);
+    static JSONRPC_STATUS Button(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

Rechi reviewed (commented):
Please follow the [code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).
You can also use `git-clang-format` or `clang-format-diff` to format your changes.

rmrector commented:
Well, my initial thought was an even simpler boolean param for "long press" - useful for remote interfaces that already have their own long press action, like mobile touchscreen interfaces. Not _all_ client implementations will be easier leaving that up to Kodi, and I think it often makes sense to let the remote client decide what is "long press".

Kodi keeping state seems a lot more complicated an implementation than passing in another param, and I don't expect the JSON-RPC interface to be particularly responsive for most games anyhow.

Maybe the stateful input can be implemented as another method? JSON-RPC is stateless so all state and client management would need to go up to some other module, with JSON-RPC calls just passing along some sort of identifier for the client.

bendavid commented:
I would not like to require that the client sends separate key down and key up events.  In some cases these actions may be generated by triggers which are not actually buttons (Home Assistant automation for example), in which case a single event corresponding to a short key press makes the most sense.  Fully agreed that adding an optional parameter for holdtime with default 0 is sensible and useful.

For what concerns handling keydown and keyup events, I see two options:

1) Implement a separate RPC call for each event type

2) Change the "Input.Button" call to "Input.Event" and add a required parameter for the event type
(this implies also/eventually supporting the other event types, but means this call will end up with a large set of optional parameters corresponding to the full set of supported event types.)

Either way JSON-RPC stays stateless, and all state is handled by the InputManager.

Any strong preferences between the two solutions?

rmrector commented:
I suggest two methods splitting on the need for an explicit client: one like you have now, when we just need to push a button, and another when there is an explicit client context that can manage the keydown and keyup events, and in the future other client things, like names, player #, and client customizations.

bendavid commented:
I've renamed the method to "Input.ButtonEvent" in anticipation of adding Input.KeyDownEvent and Input.KeyUpEvent later (these have different mappings which I didn't have a chance to work out yet).

I've added an optional holdtime parameter to ButtonEvent as discussed in any case.

rmrector commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -2342,6 +2342,18 @@
+    "permission": "Navigate",
+    "params": [
+      { "name": "button", "type": "string", "required": true, "description": "Button name" },
+      { "name": "keymap", "type": "string", "required": true, "description": "Keymap name (KB, XG, R1, or R2)" },
```
Is the list of possible "keymap" values static enough that it could be declared with an `enum`?
```suggestion
      { "name": "keymap", "type": "string", "required": true, "description": "Keymap name (KB, XG, R1, or R2)", "enum": [ "KB", "XG", "R1", "R2" ] },
```

bendavid replied:
I think supporting the universal remote map makes this not feasible, since that one takes button strings of the form "obc<number>".

For the others it should be pretty static, but the list is quite long (probably 150-200 in total)

bendavid replied:
Ah, sorry, I thought you meant the buttons rather than the keymap.  The list of keymaps is indeed static and can be an enum as you suggest.

rmrector commented:
Do details on what the possible "keymap" values mean exist somewhere other than this function? If not, can they be added to the description?

Other than those two small things the JSON-RPC implementation looks good to me.

bendavid commented:
I believe I found the description myself from here:
https://github.com/xbmc/xbmc/blob/master/tools/EventClients/lib/python/xbmcclient.py#L335-L344

Rechi commented on xbmc/interfaces/json-rpc/InputOperations.h:
```diff
@@ -19,6 +19,7 @@ namespace JSONRPC
   public:
     static JSONRPC_STATUS SendText(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result);
     static JSONRPC_STATUS ExecuteAction(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result);
+    static JSONRPC_STATUS ButtonEvent(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

bendavid replied:
I wanted to keep this consistent with the rest of the function definitions in this file.

I can update all of them rather if preferred...

Rechi replied:
Please use the current coding style for new code. Existing code can stay as is.

bendavid commented:
Small doubt from my side:  Do I need to worry about accessing the InputManager directly rather than going through the ApplicationMessenger or similar (e.g. for thread safety reasons)?

bendavid commented:
I've updated this to use the ApplicationMessenger in any case.

garbear commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +86,32 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+{
+  std::string button = parameterObject["button"].asString();
+  std::string keymap = parameterObject["keymap"].asString();
+  int holdtime = (int)parameterObject["holdtime"].asInteger();
```
Sorry for another style request - please use the c++ version, `static_cast<int>()`

bendavid replied:
In fact this cast wasn't needed at all.

garbear commented:
I'm pleased with the logic, and handling the event off-thread is a good choice (OnKey() is re-entrant but I doubt it's thread-safe).

I have a request: Can you change the event `ButtonEvent` to `KeyEvent`? I want to distinguish between stateless buttons (with an "up" event) and stateful keys (with only a "down" event, and optional hold time parameter). The reason is so we can add `ButtonEvents` in the future that handle button up, enabling our more advanced input processing system to handle the input.

bendavid commented:
The naming is intended to follow the event types in https://github.com/xbmc/xbmc/blob/5d715eabbc0ee5409568abbef6c548a0c41e0bf2/xbmc/windowing/XBMC_events.h

As such, I had envisioned that supporting the stateful events would be done with separate "KeyUpEvent" and "KeyDownEvent" calls.

garbear commented:
Files in the `windowing/` folder are for low-level OS interaction. Our API here should be based on the higher-level input abstractions used by Kodi IMO

bendavid commented:
Ok, sure, but why invent a new naming scheme when the events we're sending by construction map 1:1 onto the types already enumerated there?
(Unless this already exists somewhere else and I missed it.)

bendavid commented:
rebased

garbear commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -84,6 +86,32 @@ JSONRPC_STATUS CInputOperations::ExecuteAction(const std::string &method, ITrans
+{
+  std::string button = parameterObject["button"].asString();
+  std::string keymap = parameterObject["keymap"].asString();
+  int holdtime = parameterObject["holdtime"].asInteger();
```
To avoid adding a compiler warning on windows, can you do:

```c++
int holdtime = static_cast<int>(parameterObject["holdtime"].asInteger());
```

and then validate that holdtime isn't < 0 to avoid wraparound when assigning to `newEvent->keybutton.holdtime`

garbear reviewed (changes requested):
Can you sanitize holdtime to avoid wraparound?

garbear commented on xbmc/interfaces/json-rpc/InputOperations.cpp:
```diff
@@ -13,6 +13,8 @@
 #include "guilib/GUIWindow.h"
 #include "guilib/GUIWindowManager.h"
+#include "input/ButtonTranslator.h"
+#include "input/InputManager.h"
```
Is this include still used?

bendavid replied:
No this is indeed unused and has been removed.

garbear commented:
I think I found an unused include. I'll merge once we figure that out.

garbear commented:
Oh, and I noticed the commits need to be squashed.

bendavid commented:
Removed unused include and squashed commits.

bendavid commented:
The build failure looks like some issue with git revisions in the test, any suggestions?

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

bendavid commented:
Any remaining issues with this?  (The merge checks look related to labels which need to be added by someone other than myself.)

garbear commented:
thanks for your addition!
